### PR TITLE
docs: hacking guide & flowchart

### PR DESCRIPTION
**Pull request by Derisis13**
# Pull Request Template

## Type of change

- [ ] Bug fix
- [ ] Feature
- [x] Documentation update

## Description

Medium-sized writeup about how to modify ani-cli for sites with similar structure.
No code was modified.

## Checklist

- [ ] any anime playing
- [ ] bumped version
---
- [ ] next, prev and replay work
- [ ] `-c` history and continue work
- [ ] `-d` downloads work
- [ ] `-s` syncplay works
- [ ] `-q` quality works
- [ ] `-v` vlc works
- [ ] `-e` select episode works
- [ ] `-S` select index works
- [ ] `-r` range selection works
- [ ] `--dub` and regular (sub) mode both work
- [ ] all providers return links (not necessarily on a single anime, use debug mode to confirm)
---
- [ ] `-h` help info is up to date
- [ ] Readme is up to date
- [ ] Man page is up to date

## Additional Testcases

- The safe bet: One Piece
- Episode 0: Saenai Heroine no Sodatekata ♭
- Unicode: Saenai Heroine no Sodatekata ♭
- Non-whole episodes: Tensei shitara slime datta ken (ep. 24.5, ep. 24.9)

### CODE REVIEW

**Comment by Derisis13** [comment]
lol svg is text format after all, so it works really well with git... I didn't write a 771 line documentation (but maybe I should bump it 6 lines?)

**Review by port19x — CHANGES_REQUESTED** [review]
Please use a rendered, non-transparent png or jpg image to embed into hacking.md

As is, the flowchart is barely visible in githubs dark mode

File: hacking.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Hacking ani-cli
+Ani-cli is set up to scrape one platform - currently allanime. Supporting multiple sources at a time would require more changes than we (the maintainers) find worth doing, for this reason any feature request asking for a new site is rejected.
+
+However ani-cli being open-source and the pirate anime streaming sites being so similar you can hack ani-cli to support any site that follows a few conventions.
+
+## Prequisites
+Here's the of skills you'll need and the guide will take for granted:
+- basic shell scripting
+- understanding of http(s) requests and proficiency with curl
```
Comment (port19x) [inline hacking.md:9]:
```suggestion
- understanding of http(s) requests
```

File: hacking.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+4. download player(s) for that id+episode number combination, extract links
+5. quality selection selects one that is played
+From here 1-4 need to be changed to support another site. #Reverse-engineering will answer how.
+
+## Reverse-engineering
+Many sites have various protections against reverse-engineering.
+The extension webapi-blocker can help you with bringing up the debugger that we'll use during this guide or to conceal the presence of a debugger.
+These reverse-engineering protections are always evolving though so there's no silver bullet - you'll have to do your own research on how to get around them.
+
+As a nuclear option you can resort to [Burp Suite](https://portswigger.net/burp/communitydownload).
```
Comment (justchokingaround) [inline hacking.md:32]: this is not beginner friendly, and unnecessarily complicated imo

**Comment by justchokingaround** [comment]
i've been thinking of writing a shellscript webscraping guide as a separate repo. what would be better, incorporating a lot of new information into this ani-cli guide, or having a more detailed guide in a separate repo, and have this guide only mention which things should be changed in the ani-cli script specifically?
i'll probably also add an example for a new provider implementation

**Comment by justchokingaround** [comment]
we should also mention the existence of debug mode in `sh`, since that would make the troubleshooting process much easier, rather than seeing error messages such as `Episode not found`

**Comment by port19x** [comment]
> we should also mention the existence of debug mode in `sh`, since that would make the troubleshooting process much easier, rather than seeing error messages such as `Episode not found`

Yeah, `sh -x` is big

**Comment by CoolnsX** [comment]
> we should also mention the existence of debug mode in `sh`, since that would make the troubleshooting process much easier, rather than seeing error messages such as `Episode not found`

My french brother is alive

**Comment by port19x** [comment]
> i've been thinking of writing a shellscript webscraping guide as a separate repo. what would be better, incorporating a lot of new information into this ani-cli guide, or having a more detailed guide in a separate repo, and have this guide only mention which things should be changed in the ani-cli script specifically? i'll probably also add an example for a new provider implementation

There is something like really good in your stars list iirc. I wouldn't bother

**Review by port19x — APPROVED** [review]

**Comment by justchokingaround** [comment]
blatzar/scraping-tutorial ?

**Pull request merged**